### PR TITLE
Remove dead code

### LINKED ISSUE #1171: Remove unused code
Checking #1169 revealed there to be unused components in the codebase. I ran `npx knip` on the frontend project, and it revealed some other components as well. We should see into which of these can be deleted. (the devDependencies seem false flags to me)

---
# Knip report

## Unused files (25)

* coverage-ts/assets/source-file.js
* src/__mocks__/fileMock.js
* src/__mocks__/styleMock.js
* src/__mocks__/svgMock.js
* src/components/Diff/index.ts
* src/components/FieldSet.tsx
* src/components/ImageInput.tsx
* src/components/Layout.tsx
* src/components/Modal.tsx
* src/components/PageTitle.tsx
* src/components/Portal.tsx
* src/components/ProjectMembers.tsx
* src/components/ProjectSettings.tsx
* src/components/SearchBox.tsx
* src/components/StringInput.tsx
* src/components/UnderlineNav.tsx
* src/lib/training.ts
* src/lib/useEntity.ts

## Unused dependencies (2)

| Name                 | Location     | Severity |
| :------------------- | :----------- | :------- |
| @ducanh2912/next-pwa | package.json | error    |
| react-modal          | package.json | error    |

## Unused devDependencies (14)

| Name                      | Location     | Severity |
| :------------------------ | :----------- | :------- |
| eslint-plugin-react-hooks | package.json | error    |
| @next/eslint-plugin-next  | package.json | error    |
| eslint-formatter-rdjson   | package.json | error    |
| eslint-plugin-react       | package.json | error    |
| @types/react-modal        | package.json | error    |
| identity-obj-proxy        | package.json | error    |
| umd-compat-loader         | package.json | error    |
| postcss-loader            | package.json | error    |
| @svgr/webpack             | package.json | error    |
| style-loader              | package.json | error    |
| @types/diff               | package.json | error    |
| sass-loader               | package.json | error    |
| css-loader                | package.json | error    |
| webpack                   | package.json | error    |

## Unused exports (33)

| Name                   | Location                                               | Severity |
| :--------------------- | :----------------------------------------------------- | :------- |
| ContributorItem        | src/app/(navfooter)/credits/ContributorsList.tsx:38:17 | error    |
| LibrariesEditor        | src/components/compiler/CompilerOpts.tsx:382:17        | error    |
| DiffOptsEditor         | src/components/compiler/CompilerOpts.tsx:355:17        | error    |
| OptsEditor             | src/components/compiler/CompilerOpts.tsx:306:17        | error    |
| extension              | src/lib/codemirror/default-theme.ts:134:14             | error    |
| SITE_DESCRIPTION       | src/app/(navfooter)/WelcomeInfo.tsx:12:14              | error    |
| highlightStyle         | src/lib/codemirror/default-theme.ts:98:14              | error    |
| getUserHtmlUrl         | src/components/user/UserMention.tsx:21:17              | error    |
| isMentionable          | src/components/user/UserMention.tsx:33:17              | error    |
| theme                  | src/lib/codemirror/default-theme.ts:13:14              | error    |
| COLOR_SCHEME_PRESETS   | src/lib/codemirror/color-scheme.ts:27:14               | error    |
| default                | src/lib/codemirror/color-scheme.ts:137:8               | error    |
| SearchItem             | src/components/VerticalMenu.tsx:122:17                 | error    |
| visitLayout            | src/components/CustomLayout.tsx:31:17                  | error    |
| getMatchPercentString  | src/components/ScratchList.tsx:72:17                   | error    |
| LoadedScratchList      | src/components/ScratchList.tsx:64:17                   | error    |
| ScratchItem            | src/components/ScratchList.tsx:82:17                   | error    |
| useTranslateKeys       | src/components/Shortcut.tsx:63:17                      | error    |
| useTranslateKey        | src/components/Shortcut.tsx:42:17                      | error    |
| useGetSeparator        | src/components/Shortcut.tsx:57:17                      | error    |
| KeyMap                 | src/components/Shortcut.tsx:24:14                      | error    |
| cppLanguage            | src/lib/codemirror/cpp.ts:9:14                         | error    |
| put                    | src/lib/api/request.ts:129:23                          | error    |
| errorHandledFetchJson  | src/lib/api/request.ts:50:23                           | error    |
| projectMemberUrl       | src/lib/api/urls.ts:32:17                              | error    |
| useThemeVariable       | src/lib/hooks.ts:73:17                                 | error    |
| getUserAgent           | src/lib/device.ts:5:17                                 | error    |
| useIsMounted           | src/lib/hooks.ts:92:17                                 | error    |
| isServer               | src/lib/device.ts:1:17                                 | error    |
| useIsUserProjectMember | src/lib/api.ts:409:17                                  | error    |
| useProjectMembers      | src/lib/api.ts:385:17                                  | error    |
| forkScratch            | src/lib/api.ts:119:23                                  | error    |
| useSavedScratch        | src/lib/api.ts:58:17                                   | error    |

## Unused exported types (56)

| Name            | Location                                              | Severity |
| :-------------- | :---------------------------------------------------- | :------- |
| Contributor     | src/app/(navfooter)/credits/ContributorsList.tsx:9:13 | error    |
| Props           | src/components/PlatformSelect/PlatformIcon.tsx:37:13  | error    |
| Props           | src/components/PlatformSelect/PlatformName.tsx:7:13   | error    |
| Props           | src/app/(navfooter)/settings/SliderField.tsx:11:13    | error    |
| Option          | src/app/(navfooter)/settings/RadioList.tsx:27:13      | error    |
| Props           | src/components/Scratch/ScratchToolbar.tsx:253:13      | error    |
| Props           | src/app/(navfooter)/settings/RadioList.tsx:33:13      | error    |
| CompilerOptsT   | src/components/compiler/CompilerOpts.tsx:156:13       | error    |
| Props           | src/components/Scratch/DecompilePanel.tsx:17:13       | error    |
| Props           | src/components/compiler/CompilerOpts.tsx:164:13       | error    |
| Props           | src/app/(navfooter)/settings/TextField.tsx:5:13       | error    |
| ProblemState    | src/components/Diff/CompilationPanel.tsx:25:13        | error    |
| PerSaveObj      | src/components/Diff/CompilationPanel.tsx:31:13        | error    |
| Props           | src/components/Diff/CompilationPanel.tsx:35:13        | error    |
| Props           | src/app/(navfooter)/settings/Checkbox.tsx:3:13        | error    |
| Props           | src/app/(navfooter)/settings/NavItem.tsx:11:13        | error    |
| Props           | src/app/scratch/[slug]/ScratchEditor.tsx:79:18        | error    |
| Props           | src/components/Scratch/AboutScratch.tsx:45:13         | error    |
| Props           | src/app/(navfooter)/credits/LinkList.tsx:3:13         | error    |
| Props           | src/app/(navfooter)/settings/Section.tsx:3:13         | error    |
| Props           | src/components/PlatformSelect/index.ts:3:13           | error    |
| Props           | src/components/user/AnonymousFrog.tsx:11:13           | error    |
| Props           | src/components/ColorSchemeEditor.tsx:50:18            | error    |
| Props           | src/components/Editor/CodeMirror.tsx:34:18            | error    |
| Props           | src/components/ColorSchemePicker.tsx:7:18             | error    |
| ColorPresetName | src/lib/codemirror/color-scheme.ts:98:13              | error    |
| Colors          | src/lib/codemirror/color-scheme.ts:25:13              | error    |
| Props           | src/components/user/UserAvatar.tsx:15:13              | error    |
| Props           | src/components/user/UserMention.tsx:9:13              | error    |
| Props           | src/components/YourScratchList.tsx:7:13               | error    |
| VerticalSplit   | src/components/CustomLayout.tsx:14:18                 | error    |
| Props           | src/components/CustomLayout.tsx:51:18                 | error    |
| Props           | src/components/ErrorBoundary.tsx:9:18                 | error    |
| Props           | src/components/VerticalMenu.tsx:20:18                 | error    |
| Props           | src/components/user/UserLink.tsx:7:13                 | error    |
| Pane            | src/components/CustomLayout.tsx:21:18                 | error    |
| HorizontalSplit | src/components/CustomLayout.tsx:7:18                  | error    |
| Props           | src/components/AsyncButton.tsx:13:18                  | error    |
| Props           | src/components/PlatformLink.tsx:8:13                  | error    |
| Props           | src/components/Diff/DragBar.tsx:5:18                  | error    |
| Props           | src/components/Scratch/index.ts:3:13                  | error    |
| Props           | src/components/Breadcrumbs.tsx:9:18                   | error    |
| Props           | src/components/GhostButton.tsx:7:13                   | error    |
| Props           | src/components/NumberInput.tsx:7:13                   | error    |
| Props           | src/components/ScoreBadge.tsx:48:13                   | error    |
| Props           | src/components/ThemePicker.tsx:9:18                   | error    |
| Props           | src/components/Diff/Diff.tsx:204:13                   | error    |
| Props           | src/components/Nav/Nav.tsx:18:18                      | error    |
| Section         | src/lib/i18n/translate.ts:11:13                       | error    |
| Props           | src/components/Select2.tsx:7:13                       | error    |
| Props           | src/components/Select.tsx:7:13                        | error    |
| TabContent      | src/components/Tabs.tsx:33:13                         | error    |
| TabProps        | src/components/Tabs.tsx:35:13                         | error    |
| Props           | src/components/Tabs.tsx:95:13                         | error    |
| DiffHeader      | src/lib/api/types.ts:84:13                            | error    |
| Stats           | src/lib/api.ts:366:18                                 | error    |

## Duplicate exports (2)

| Name                         | Location                            | Severity |
| :--------------------------- | :---------------------------------- | :------- |
| extension|default            | src/lib/codemirror/default-theme.ts | error    |
| COLOR_SCHEME_PRESETS|default | src/lib/codemirror/color-scheme.ts  | error    |

### PR DESCRIPTION
**Pull request by marijnvdwerf**
Probably requires some more discussion

closes #1171

### CODE REVIEW

**Review by ethteck — APPROVED** [review]

File: frontend/src/lib/codemirror/languageServer.ts [hunk 1]
```diff
@@ -322,10 +322,6 @@ class LanguageServerPlugin implements PluginValue {
         }
     }
 
-    requestDiagnostics(view: EditorView) {
```
Comment (bates64) [inline frontend/src/lib/codemirror/languageServer.ts:325]: I reckon we should keep this, ideally clangd provides live diagnostics in the near future

File: frontend/src/lib/hooks.ts [hunk 2]
```diff
@@ -70,35 +70,10 @@ export function useWarnBeforeUnload(enabled: boolean, message = "Are you sure yo
...
-
 export function usePageTitle(...breadcrumbs: string[]) {
     const title = joinTitles(...breadcrumbs)
 
     useEffect(() => {
         document.title = title
     }, [title])
 }
-
-export function useIsMounted() {
```
Comment (bates64) [inline frontend/src/lib/hooks.ts:92]: IMO we should keep this as its commonly needed, even if we don't use it right now

File: frontend/src/lib/useEntity.ts [hunk 3]
```diff
@@ -1,68 +0,0 @@
...
-                setLocalPatch({ ...localPatch, ...partial })
-            } else {
-                setLocalPatch(partial)
-            }
-        },
-        reset() {
-            setLocalPatch(null)
-        },
-    }]
-}
```
Comment (bates64) [inline frontend/src/lib/useEntity.ts:68]: Shame we didn't use this in more places

**Review by bates64 — COMMENTED** [review]
Couple comments.
Worth remembering this PR in future if we need Portal or the mocks back

**Pull request merged**